tests/algorithms/test_cg.py

```diff
@@ -0,0 +1,154 @@
+"""Tests for the conjugate gradient method."""
+
+# Copyright 2024 Physikalisch-Technische Bundesanstalt
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#       http://www.apache.org/licenses/LICENSE-2.0
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import pytest
+import scipy
+import torch
+from scipy.sparse.linalg import cg as cg_scp
+
+from mrpro.algorithms import conjugate_gradient
+from mrpro.operators import MatrixOp
+from tests import RandomGenerator
+
+
+@pytest.fixture(
+    params=[
+        (1, 32, True),  # (batch-size, vector-size, flag for complex-valued system or not)
+        (1, 32, False),
+        (4, 32, True),
+        (4, 32, False),
+    ]
+)
+def system(request):
+    """Generate data for creating a system Hx=b with linear and self-adjoint
+    H."""
+    random_generator = RandomGenerator(seed=0)
+
+    # get parameters for the test
+    batchsize, vectorsize, complex_valued = request.param
+
+    # if mb=1, it corresponds to one linear system; for mb>1, multiple systems
+    # are considered simultaneously
+    matrix_shape = (batchsize, vectorsize, vectorsize) if batchsize > 1 else (vectorsize, vectorsize)
+    vector_shape = (batchsize, vectorsize, 1) if batchsize > 1 else (vectorsize,)
+    a_matrix = (
+        random_generator.complex64_tensor(size=matrix_shape, high=1.0)
+        if complex_valued
+        else random_generator.float32_tensor(size=matrix_shape, low=-1.0, high=1.0)
+    )
+
+    # make sure H is self-adjoint
+    h_matrix = a_matrix.mH @ a_matrix
+
+    # construct matrix multiplication as LinearOperator
+    h_operator = MatrixOp(h_matrix)
+
+    # create ground-truth data and right-hand side of the system
+    vector = (
+        random_generator.complex64_tensor(
+            size=vector_shape,
+            high=1.0,
+        )
+        if complex_valued
+        else random_generator.float32_tensor(
+            size=vector_shape,
+            low=-1.0,
+            high=1.0,
+        )
+    )
+    (right_hand_side,) = h_operator(vector)
+
+    return h_operator, right_hand_side, vector
+
+
+def test_cg_convergence(system):
+    """Test if CG delivers accurate solution."""
+
+    # create operator, right-hand side and ground-truth data
+    h_operator, right_hand_side, solution = system
```

Comment:
question: why is h_operator and  right_hand_side? As in, why is the operator named like in the equation Hx=b, b and x are not..

h_operator and b
or linear_op and right_hand_side?